File: leetcode/5.最长回文子串.js
/*
 * @lc app=leetcode.cn id=5 lang=javascript
 *
 * [5] 最长回文子串
 */
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    if (!s) return '';
    if (s.length < 2) return s;
    s = '^#' + s.split('').join('#') + '#$';
    var radius = new Array(s.length).fill(0);
    var C = 0,
        centerIndex = 0,
        maxRight = 0,
        maxLen = 0;

    for (var i = 1; i < s.length - 1; i++) {
        // 计算初始回文半径, i' = 2 * C - i
        radius[i] = maxRight > i ? Math.min(maxRight - i, radius[2 * C - i]) : 0;
        // 扩展半径
        while (
            s[i + 1 + radius[i]] &&
            s[i - 1 - radius[i]] &&
            s[i + 1 + radius[i]] === s[i - 1 - radius[i]]
        )
            radius[i]++;
        // 更新当前搜索的最大右边界和位置
        if (i + radius[i] > maxRight) {
            C = i;
            maxRight = i + radius[i];
        }
        // 更新最大回文串长度及位置
        if (maxLen < radius[i]) {
            maxLen = radius[i];
            centerIndex = i;
        }
    }

    return s
        .slice(centerIndex - maxLen, centerIndex + maxLen + 1)
        .split('#')
        .join('');
};

File: leetcode/5.最长回文子串.py
#
# @lc app=leetcode.cn id=5 lang=python3
#
# [5] 最长回文子串
#


class Solution:
    def longestPalindrome(self, s: str) -> str:
        # 预处理
        s = '#'+'#'.join(s)+'#'

        RL = [0]*len(s)
        MaxRight = 0
        pos = 0
        MaxLen = 0
        for i in range(len(s)):
            if i < MaxRight:
                RL[i] = min(RL[2*pos-i], MaxRight-i)
            else:
                RL[i] = 1
            # 尝试扩展，注意处理边界
            while i-RL[i] >= 0 and i+RL[i] < len(s) and s[i-RL[i]] == s[i+RL[i]]:
                RL[i] += 1
            # 更新MaxRight,pos
            if RL[i]+i-1 > MaxRight:
                MaxRight = RL[i]+i-1
                pos = i
            # 更新最长回文串的长度
            MaxLen = max(MaxLen, RL[i])
        return MaxLen-1
